gapic-common/lib/gapic/logging_concerns.rb

```diff
@@ -0,0 +1,211 @@
+
+        attr_reader :service
+
+        attr_reader :endpoint
+
+        attr_accessor :message
+
+        attr_writer :source_location
+
+        attr_reader :fields
+
+        def set name, value
+          fields[name] = value
+        end
+
+        def set_system_name
+          set "system", system_name
+        end
+
+        def set_service
+          set "serviceName", service
+        end
+
+        def set_credentials_fields creds
+          creds = creds.client if creds.respond_to? :client
+          set "credentialsId", creds.object_id
+          set "credentialsType", creds.class.name
+          set "credentialsScope", creds.scope if creds.respond_to? :scope
+          set "useSelfSignedJWT", creds.enable_self_signed_jwt? if creds.respond_to? :enable_self_signed_jwt?
+          set "universeDomain", creds.universe_domain if creds.respond_to? :universe_domain
+        end
+
+        def source_location
+          @source_location ||= Google::Logging::SourceLocation.for_caller omit_files: OMIT_FILES,
+                                                                          locations: @caller_locations
+        end
+
+        def set_source_location_from locations
+          @caller_locations = locations
+          @source_location = nil
+        end
+
+        # @private
+        def build
+          Google::Logging::Message.from message: message, source_location: source_location, fields: fields
+        end
+      end
+    end
+
+    ##
+    # @private
+    # Initialize logging concerns.
+    #
+    def setup_logging logger: :default,
+                      stream: nil,
+                      formatter: nil,
+                      level: nil,
+                      system_name: nil,
+                      service: nil,
+                      endpoint: nil,
+                      client_id: nil
+      service = LoggingConcerns.normalize_service service
+      system_name = LoggingConcerns.normalize_system_name system_name
+      logging_env = ENV["GOOGLE_SDK_RUBY_LOGGING_GEMS"].to_s.downcase
+      logger = nil if ["false", "none"].include? logging_env
+      if logger == :default
+        logger = nil
+        if ["true", "all"].include?(logging_env) || logging_env.split(",").include?(system_name)
+          stream ||= $stderr
+          level ||= "DEBUG"
+          formatter ||= Google::Logging::StructuredFormatter.new if Google::Cloud::Env.get.logging_agent_expected?
+          logger = Logger.new stream, progname: system_name, level: level, formatter: formatter
+        end
+      end
+      @logger = logger
+      @stub_logger = StubLogger.new logger: logger,
+                                    system_name: system_name,
+                                    service: service,
+                                    endpoint: endpoint,
+                                    client_id: client_id
+    end
+
+    # @private
+    attr_reader :stub_logger
+
+    class << self
+      # @private
+      def random_uuid4
```

Comment:
Main reason I chose to avoid SecureRandom here is that its internal implementation can block for brief periods of time (e.g. if `/dev/urandom` is being used), and for logging purposes we have a fairly strict requirement against that. That said, there is `Random::Formatter` (i.e. allowing `Random.uuid`) which I _would_ use except it seems to require Ruby 3.1, and we're not quite scheduled to drop 3.0 support yet until next spring.